File: example/erlang/src/hello_nif.erl
-module(hello_nif).
-export([add/2]).
-on_load(init/0).

-define(APPNAME, hello).
-define(LIBNAME, hello).

add(_, _) ->
    not_loaded(?LINE).

init() ->
    PrivDir = code:priv_dir(?APPNAME),
    PrivLibDir = filename:join(PrivDir, lib),
    ok = 
        case os:type() of
            {win32, _} ->
                dll_loader_helper:add_dll_directory(PrivLibDir);
            _ ->
                ok
        end,
    SoName = filename:join(PrivDir, hello_nif),
    erlang:load_nif(SoName, 0).

not_loaded(Line) ->
    erlang:nif_error({not_loaded, [{module, ?MODULE}, {line, Line}]}).

File: dll_loader_helper/example/elixir/lib/hello_nif.ex
defmodule :hello_nif do
  @on_load :load_nif
  def load_nif do
    :ok =
      case :os.type do
        {:win32, _} -> DLLLoaderHelper.addDLLDirectory("#{:code.priv_dir(:hello)}/lib")
        _ -> :ok
      end
    nif_file = '#{:code.priv_dir(:hello)}/hello_nif'

    case :erlang.load_nif(nif_file, 0) do
      :ok -> :ok
      {:error, {:reload, _}} -> :ok
      {:error, reason} -> IO.puts("Failed to load nif: #{inspect(reason)}")
    end
  end

  def add(_a, _b), do: :erlang.nif_error(:not_loaded)
end
